### PR TITLE
Font Library: return null if a font collection is not registered

### PR DESCRIPTION
**Pull request by matiasbenedetto**
## What?
Font Library: return null if a font collection is not registered

## Why?
- improves code quality
- address feedback from: https://github.com/WordPress/wordpress-develop/pull/6034#discussion_r1479440956

## How?
returning null when a font collection is not registered

## Testing Instructions
- Run unit tests

### CODE REVIEW

**Review by creativecoder — COMMENTED** [review]
This looks good to me. I see other Core functions vary on whether they return `null` or `false` in this case, but I think returning null seems most appropriate, as you tried to get something, but there's nothing there.

I see that phpunit tests are failing, I think because the `WP_Font_Library` class has already been committed to WordPress trunk, that class is overriding here, and it doesn't yet have the updated method that returns null? @youknowriad @getdave does that seem right? Any advice on how to get CI passing?

**Comment by youknowriad** [comment]
> I see that phpunit tests are failing, I think because the WP_Font_Library class has already been committed to WordPress trunk, that class is overriding here, and it doesn't yet have the updated method that returns null?

I think at this point, we could consider removing the phpunit tests from Gutenberg, because from now on, they will be maintained on Core. That's how we approached this kind of things in the past.

File: lib/compat/wordpress-6.5/fonts/class-wp-font-library.php [hunk 1]
```diff
@@ -117,14 +117,14 @@ public function get_font_collections() {
 		 * @since 6.5.0
 		 *
 		 * @param string $slug Font collection slug.
-		 * @return WP_Font_Collection|WP_Error Font collection object,
-		 *                                     or WP_Error object if the font collection doesn't exist.
+		 * @return WP_Font_Collection|null Font collection object,
+		 *                                     or null object if the font collection doesn't exist.
```
Comment (creativecoder) [inline lib/compat/wordpress-6.5/fonts/class-wp-font-library.php]:
```suggestion
		 *                                     or null if the font collection doesn't exist.
```

File: lib/compat/wordpress-6.5/fonts/class-wp-rest-font-collections-controller.php [hunk 2]
```diff
@@ -144,10 +144,8 @@ public function get_item( $request ) {
 			$slug       = $request->get_param( 'slug' );
 			$collection = WP_Font_Library::get_instance()->get_font_collection( $slug );
 
-			// If the collection doesn't exist returns a 404.
-			if ( is_wp_error( $collection ) ) {
-				$collection->add_data( array( 'status' => 404 ) );
-				return $collection;
+			if ( ! $collection ) {
+				return new WP_Error( 'font_collection_not_found', __( 'Font collection not found.' ), array( 'status' => 404 ) );
```
Comment (creativecoder) [inline lib/compat/wordpress-6.5/fonts/class-wp-rest-font-collections-controller.php]:
```suggestion
				return new WP_Error( 'rest_font_collection_not_found', __( 'Font collection not found.' ), array( 'status' => 404 ) );
```

**Comment by matiasbenedetto** [comment]
A PR removing the tests from the already merged classes: https://github.com/WordPress/gutenberg/pull/58752 that we would need to merge before this one.

**Comment by creativecoder** [comment]
Do we need to remove all tests right now?

Gutenberg supports the current and previous release of WordPress, so having tests that cover the 6.5 compat code seems to have some value until 6.4 isn't supported any longer.

What do you think of just removing that one test, for now, until WP trunk is in sync with these changes?

**Comment by youknowriad** [comment]
>Gutenberg supports the current and previous release of WordPress, so having tests that cover the 6.5 compat code seems to have some value until 6.4 isn't supported any longer.

you're right, I think the compat tests should stay in Gutenberg because the code they cover will remain in Gutenberg but for the classes that moved to Core, we should remove their tests.

**Comment by matiasbenedetto** [comment]
> What do you think of just removing that one test, for now, until WP trunk is in sync with these changes?

Removing just some tests makes things difficult to track and could give us confidence that things are being appropriately tested when they could not. 


> you're right, I think the compat tests should stay in Gutenberg because the code they cover will remain in Gutenberg

In that case we would we need to find a way to run the tests loading the classes tested from Gutenberg and not from core. I'm not sure how to do that.

**Comment by matiasbenedetto** [comment]
> In that case we would we need to find a way to run the tests loading the classes tested from Gutenberg and not from core. I'm not sure how to do that.

Should we add a `_Gutenberg` suffix to the name of the classes that were already merged to core to be able to test them? Example: `WP_Font_Library_Gutenberg`

**Comment by creativecoder** [comment]
> I think the compat tests should stay in Gutenberg because the code they cover will remain in Gutenberg

I think @youknowriad might be referring to font code and tests that aren't being ported to Core. The only thing I'm aware of are the fonts bc-layer code, as well as `gutenberg_convert_legacy_font_family_format` and the related test file. Doesn't look like those are being removed, so this shouldn't be a problem.

**Comment by youknowriad** [comment]
@coderkevin Yes, I was indeed referring to these files :) 

Now for testing the ones that are already on Core, to be honest I don't see a value. What would be good would be to check that they're kept in sync with core classes but we don't have any of that so far and it's not specific to font library.

**Comment by creativecoder** [comment]
I've merged https://github.com/WordPress/gutenberg/pull/58752 and have rebased this branch, waiting for CI to pass.

**Comment by creativecoder** [comment]
There's still some weirdness here, even with removing the phpunit tests already committed to Core.

`WP_Font_Library::get_font_collection` returns WP_Error on current WP trunk, but null here. We need to account for both in the controller, since unit tests are run against WP trunk.

For now I've added a check for a WP_Error instance in the controller, which we should remove once WP trunk is updated with the code in this PR.

**Review by creativecoder — APPROVED** [review]
CI is finally passing, so this is now ready to merge.

**Pull request merged**